File: packages/mcp-server/src/classification/heuristics.ts
/**
 * Heuristic-based node type classification
 * Fast keyword matching without LLM calls
 */
import type { NodeType } from '../storage/types.js';

export interface ClassificationResult {
  readonly type: NodeType;
  readonly confidence: number;
  readonly reasoning: string;
}

interface Pattern {
  readonly keywords: ReadonlyArray<string>;
  readonly weight: number;
}

const CLASSIFICATION_PATTERNS: Record<NodeType, ReadonlyArray<Pattern>> = {
  decision: [
    { keywords: ['chose', 'choose', 'decided', 'decide', 'selected', 'select'], weight: 1.0 },
    { keywords: ['decision', 'choice', 'selection'], weight: 0.9 },
    { keywords: ['adopt', 'use', 'switch to', 'move to'], weight: 0.8 },
    { keywords: ['instead of', 'over', 'rather than', 'vs'], weight: 0.7 },
    { keywords: ['because', 'since', 'reason', 'rationale'], weight: 0.6 },
    { keywords: ['alternative', 'option', 'considered'], weight: 0.5 }
  ],

  component: [
    { keywords: ['service', 'module', 'class', 'function', 'method'], weight: 1.0 },
    { keywords: ['component', 'controller', 'model', 'view'], weight: 0.9 },
    { keywords: ['handles', 'manages', 'implements', 'provides'], weight: 0.8 },
    { keywords: ['api', 'endpoint', 'route', 'handler'], weight: 0.7 },
    { keywords: ['created', 'built', 'implemented'], weight: 0.6 }
  ],

  convention: [
    { keywords: ['always', 'never', 'must', 'should', 'shall'], weight: 1.0 },
    { keywords: ['convention', 'standard', 'practice', 'guideline'], weight: 0.9 },
    { keywords: ['rule', 'policy', 'requirement'], weight: 0.8 },
    { keywords: ['style', 'format', 'naming', 'pattern'], weight: 0.7 },
    { keywords: ['all', 'every', 'each', 'any'], weight: 0.6 }
  ],

  concept: [
    { keywords: ['concept', 'idea', 'notion', 'model'], weight: 1.0 },
    { keywords: ['represents', 'means', 'refers to', 'is'], weight: 0.9 },
    { keywords: ['lifecycle', 'workflow', 'process', 'flow'], weight: 0.8 },
    { keywords: ['state', 'status', 'phase', 'stage'], weight: 0.7 },
    { keywords: ['domain', 'business', 'entity'], weight: 0.6 }
  ],

  pattern: [
    { keywords: ['pattern', 'approach', 'strategy', 'technique'], weight: 1.0 },
    { keywords: ['reusable', 'generic', 'abstract', 'common'], weight: 0.9 },
    { keywords: ['design pattern', 'architectural pattern'], weight: 0.95 },
    { keywords: ['factory', 'singleton', 'observer', 'repository'], weight: 0.8 },
    { keywords: ['template', 'blueprint', 'recipe'], weight: 0.7 }
  ],

  issue: [
    { keywords: ['issue', 'problem', 'bug', 'error', 'defect'], weight: 1.0 },
    { keywords: ['broken', 'failing', 'not working'], weight: 0.9 },
    { keywords: ['debt', 'technical debt', 'todo', 'fixme'], weight: 0.8 },
    { keywords: ['bottleneck', 'performance', 'slow'], weight: 0.7 },
    { keywords: ['needs fix', 'needs refactor', 'improvement needed'], weight: 0.6 }
  ],

  session: [
    { keywords: ['session', 'worked on', 'accomplished', 'completed'], weight: 1.0 },
    { keywords: ['today', 'this session', 'summary'], weight: 0.9 },
    { keywords: ['implemented', 'fixed', 'added', 'updated'], weight: 0.7 },
    { keywords: ['progress', 'status update'], weight: 0.6 }
  ]
};

/**
 * Classify node type based on title and content
 */
export function classifyNodeType(
  title: string,
  content: string,
  sourceContext?: 'hook_write' | 'hook_commit' | 'hook_session' | 'manual'
): ClassificationResult {
  const text = `${title} ${content}`.toLowerCase();

  // Context-based boosting
  const contextBoost: Record<string, Partial<Record<NodeType, number>>> = {
    hook_write: { component: 0.3, convention: 0.2 },
    hook_commit: { decision: 0.3, issue: 0.2 },
    hook_session: { session: 0.5 },
    manual: {} // No boost for manual
  };

  const boost = contextBoost[sourceContext || 'manual'] || {};

  // Score each type
  const scores: Record<NodeType, number> = {
    decision: 0,
    component: 0,
    convention: 0,
    concept: 0,
    pattern: 0,
    issue: 0,
    session: 0
  };

  for (const [type, patterns] of Object.entries(CLASSIFICATION_PATTERNS)) {
    let typeScore = 0;

    for (const pattern of patterns) {
      for (const keyword of pattern.keywords) {
        if (text.includes(keyword)) {
          typeScore += pattern.weight;
        }
      }
    }

    // Apply context boost
    const nodeType = type as NodeType;
    const boostAmount = boost[nodeType] || 0;
    typeScore += boostAmount;

    scores[nodeType] = typeScore;
  }

  // Find best match
  let bestType: NodeType = 'concept'; // Default fallback
  let bestScore = 0;

  for (const [type, score] of Object.entries(scores)) {
    if (score > bestScore) {
      bestScore = score;
      bestType = type as NodeType;
    }
  }

  // Normalize confidence (cap at 0.95 for heuristics)
  const maxPossibleScore = 10; // Rough estimate
  const confidence = Math.min(0.95, bestScore / maxPossibleScore);

  // Generate reasoning
  const reasoning = generateReasoning(bestType, text, scores);

  return {
    type: bestType,
    confidence,
    reasoning
  };
}

/**
 * Generate human-readable reasoning for classification
 */
function generateReasoning(
  chosenType: NodeType,
  text: string,
  scores: Record<NodeType, number>
): string {
  const patterns = CLASSIFICATION_PATTERNS[chosenType];
  const matchedKeywords: string[] = [];

  for (const pattern of patterns) {
    for (const keyword of pattern.keywords) {
      if (text.includes(keyword)) {
        matchedKeywords.push(keyword);
        if (matchedKeywords.length >= 3) break;
      }
    }
    if (matchedKeywords.length >= 3) break;
  }

  const topScores = Object.entries(scores)
    .filter(([t]) => t !== chosenType)
    .sort(([, a], [, b]) => b - a)
    .slice(0, 2);

  let reasoning = `Classified as '${chosenType}' based on keywords: ${matchedKeywords.join(', ')}`;

  if (topScores.length > 0 && topScores[0] && topScores[0][1] > scores[chosenType] * 0.7) {
    reasoning += `. Also considered '${topScores[0][0]}' (score: ${topScores[0][1].toFixed(2)})`;
  }

  return reasoning;
}

/**
 * Check if classification is confident enough
 */
export function isConfidentClassification(result: ClassificationResult): boolean {
  return result.confidence >= 0.6;
}

/**
 * Suggest alternative types if confidence is low
 */
export function suggestAlternatives(
  title: string,
  content: string
): ReadonlyArray<ClassificationResult> {
  const text = `${title} ${content}`.toLowerCase();
  const scores: Array<{ type: NodeType; score: number }> = [];

  for (const [type, patterns] of Object.entries(CLASSIFICATION_PATTERNS)) {
    let typeScore = 0;

    for (const pattern of patterns) {
      for (const keyword of pattern.keywords) {
        if (text.includes(keyword)) {
          typeScore += pattern.weight;
        }
      }
    }

    if (typeScore > 0) {
      scores.push({ type: type as NodeType, score: typeScore });
    }
  }

  // Sort by score descending
  scores.sort((a, b) => b.score - a.score);

  // Return top 3
  const maxPossibleScore = 10;
  return scores.slice(0, 3).map(({ type, score }) => ({
    type,
    confidence: Math.min(0.95, score / maxPossibleScore),
    reasoning: `Score: ${score.toFixed(2)}`
  }));
}

File: internal/classification/heuristics.go
package classification

import (
	"fmt"
	"math"
	"strings"

	"github.com/Curt-Park/autology/internal/storage"
)

// ClassificationResult represents the result of node type classification
type ClassificationResult struct {
	Type       storage.NodeType
	Confidence float64
	Reasoning  string
}

// Pattern represents a keyword pattern for classification
type Pattern struct {
	Keywords []string
	Weight   float64
}

// Classification patterns for each node type
var classificationPatterns = map[storage.NodeType][]Pattern{
	storage.NodeTypeDecision: {
		{Keywords: []string{"chose", "choose", "decided", "decide", "selected", "select"}, Weight: 1.0},
		{Keywords: []string{"decision", "choice", "selection"}, Weight: 0.9},
		{Keywords: []string{"adopt", "use", "switch to", "move to"}, Weight: 0.8},
		{Keywords: []string{"instead of", "over", "rather than", "vs"}, Weight: 0.7},
		{Keywords: []string{"because", "since", "reason", "rationale"}, Weight: 0.6},
		{Keywords: []string{"alternative", "option", "considered"}, Weight: 0.5},
	},
	storage.NodeTypeComponent: {
		{Keywords: []string{"service", "module", "class", "function", "method"}, Weight: 1.0},
		{Keywords: []string{"component", "controller", "model", "view"}, Weight: 0.9},
		{Keywords: []string{"handles", "manages", "implements", "provides"}, Weight: 0.8},
		{Keywords: []string{"api", "endpoint", "route", "handler"}, Weight: 0.7},
		{Keywords: []string{"created", "built", "implemented"}, Weight: 0.6},
	},
	storage.NodeTypeConvention: {
		{Keywords: []string{"always", "never", "must", "should", "shall"}, Weight: 1.0},
		{Keywords: []string{"convention", "standard", "practice", "guideline"}, Weight: 0.9},
		{Keywords: []string{"rule", "policy", "requirement"}, Weight: 0.8},
		{Keywords: []string{"style", "format", "naming", "pattern"}, Weight: 0.7},
		{Keywords: []string{"all", "every", "each", "any"}, Weight: 0.6},
	},
	storage.NodeTypeConcept: {
		{Keywords: []string{"concept", "idea", "notion", "model"}, Weight: 1.0},
		{Keywords: []string{"represents", "means", "refers to", "is"}, Weight: 0.9},
		{Keywords: []string{"lifecycle", "workflow", "process", "flow"}, Weight: 0.8},
		{Keywords: []string{"state", "status", "phase", "stage"}, Weight: 0.7},
		{Keywords: []string{"domain", "business", "entity"}, Weight: 0.6},
	},
	storage.NodeTypePattern: {
		{Keywords: []string{"pattern", "approach", "strategy", "technique"}, Weight: 1.0},
		{Keywords: []string{"reusable", "generic", "abstract", "common"}, Weight: 0.9},
		{Keywords: []string{"design pattern", "architectural pattern"}, Weight: 0.95},
		{Keywords: []string{"factory", "singleton", "observer", "repository"}, Weight: 0.8},
		{Keywords: []string{"template", "blueprint", "recipe"}, Weight: 0.7},
	},
	storage.NodeTypeIssue: {
		{Keywords: []string{"issue", "problem", "bug", "error", "defect"}, Weight: 1.0},
		{Keywords: []string{"broken", "failing", "not working"}, Weight: 0.9},
		{Keywords: []string{"debt", "technical debt", "todo", "fixme"}, Weight: 0.8},
		{Keywords: []string{"bottleneck", "performance", "slow"}, Weight: 0.7},
		{Keywords: []string{"needs fix", "needs refactor", "improvement needed"}, Weight: 0.6},
	},
	storage.NodeTypeSession: {
		{Keywords: []string{"session", "worked on", "accomplished", "completed"}, Weight: 1.0},
		{Keywords: []string{"today", "this session", "summary"}, Weight: 0.9},
		{Keywords: []string{"implemented", "fixed", "added", "updated"}, Weight: 0.7},
		{Keywords: []string{"progress", "status update"}, Weight: 0.6},
	},
}

// Context-based score boosting
var contextBoost = map[string]map[storage.NodeType]float64{
	"hook_write": {
		storage.NodeTypeComponent:  0.3,
		storage.NodeTypeConvention: 0.2,
	},
	"hook_commit": {
		storage.NodeTypeDecision: 0.3,
		storage.NodeTypeIssue:    0.2,
	},
	"hook_session": {
		storage.NodeTypeSession: 0.5,
	},
	"manual": {},
}

// ClassifyNodeType classifies a node based on title and content
func ClassifyNodeType(title, content, sourceContext string) ClassificationResult {
	text := strings.ToLower(title + " " + content)

	// Get context boost
	boost := contextBoost[sourceContext]
	if boost == nil {
		boost = contextBoost["manual"]
	}

	// Score each type
	scores := make(map[storage.NodeType]float64)
	for _, nodeType := range storage.NodeTypes {
		scores[nodeType] = 0
	}

	for nodeType, patterns := range classificationPatterns {
		typeScore := 0.0

		for _, pattern := range patterns {
			for _, keyword := range pattern.Keywords {
				if strings.Contains(text, keyword) {
					typeScore += pattern.Weight
				}
			}
		}

		// Apply context boost
		if boostAmount, ok := boost[nodeType]; ok {
			typeScore += boostAmount
		}

		scores[nodeType] = typeScore
	}

	// Find best match
	bestType := storage.NodeTypeConcept // Default fallback
	bestScore := 0.0

	for nodeType, score := range scores {
		if score > bestScore {
			bestScore = score
			bestType = nodeType
		}
	}

	// Normalize confidence (cap at 0.95 for heuristics)
	maxPossibleScore := 10.0 // Rough estimate
	confidence := math.Min(0.95, bestScore/maxPossibleScore)

	// Generate reasoning
	reasoning := generateReasoning(bestType, text, scores)

	return ClassificationResult{
		Type:       bestType,
		Confidence: confidence,
		Reasoning:  reasoning,
	}
}

// generateReasoning creates human-readable reasoning for the classification
func generateReasoning(chosenType storage.NodeType, text string, scores map[storage.NodeType]float64) string {
	patterns := classificationPatterns[chosenType]
	matchedKeywords := make([]string, 0, 3)

	for _, pattern := range patterns {
		for _, keyword := range pattern.Keywords {
			if strings.Contains(text, keyword) {
				matchedKeywords = append(matchedKeywords, keyword)
				if len(matchedKeywords) >= 3 {
					break
				}
			}
		}
		if len(matchedKeywords) >= 3 {
			break
		}
	}

	// Find top alternative scores
	type scoreEntry struct {
		nodeType storage.NodeType
		score    float64
	}

	alternatives := make([]scoreEntry, 0)
	for nodeType, score := range scores {
		if nodeType != chosenType {
			alternatives = append(alternatives, scoreEntry{nodeType, score})
		}
	}

	// Sort alternatives by score descending
	for i := 0; i < len(alternatives)-1; i++ {
		for j := i + 1; j < len(alternatives); j++ {
			if alternatives[j].score > alternatives[i].score {
				alternatives[i], alternatives[j] = alternatives[j], alternatives[i]
			}
		}
	}

	reasoning := fmt.Sprintf("Classified as '%s' based on keywords: %s",
		chosenType, strings.Join(matchedKeywords, ", "))

	if len(alternatives) > 0 && alternatives[0].score > scores[chosenType]*0.7 {
		reasoning += fmt.Sprintf(". Also considered '%s' (score: %.2f)",
			alternatives[0].nodeType, alternatives[0].score)
	}

	return reasoning
}

// IsConfidentClassification checks if classification confidence is sufficient
func IsConfidentClassification(result ClassificationResult) bool {
	return result.Confidence >= 0.6
}

// SuggestAlternatives suggests alternative node types with their scores
func SuggestAlternatives(title, content string) []ClassificationResult {
	text := strings.ToLower(title + " " + content)

	type scoreEntry struct {
		nodeType storage.NodeType
		score    float64
	}

	scores := make([]scoreEntry, 0)

	for nodeType, patterns := range classificationPatterns {
		typeScore := 0.0

		for _, pattern := range patterns {
			for _, keyword := range pattern.Keywords {
				if strings.Contains(text, keyword) {
					typeScore += pattern.Weight
				}
			}
		}

		if typeScore > 0 {
			scores = append(scores, scoreEntry{nodeType, typeScore})
		}
	}

	// Sort by score descending
	for i := 0; i < len(scores)-1; i++ {
		for j := i + 1; j < len(scores); j++ {
			if scores[j].score > scores[i].score {
				scores[i], scores[j] = scores[j], scores[i]
			}
		}
	}

	// Return top 3
	maxPossibleScore := 10.0
	results := make([]ClassificationResult, 0, 3)

	limit := 3
	if len(scores) < limit {
		limit = len(scores)
	}

	for i := 0; i < limit; i++ {
		results = append(results, ClassificationResult{
			Type:       scores[i].nodeType,
			Confidence: math.Min(0.95, scores[i].score/maxPossibleScore),
			Reasoning:  fmt.Sprintf("Score: %.2f", scores[i].score),
		})
	}

	return results
}

// GetClassificationPatterns returns the classification patterns (for testing)
func GetClassificationPatterns() map[storage.NodeType][]Pattern {
	return classificationPatterns
}
